src/analysis/refactor_extract_region.ml

```diff
@@ -0,0 +1,479 @@
+          match expr.pexp_desc with
+          | Pexp_construct
+              ( ident,
+                Some
+                  { pexp_desc =
+                      Pexp_tuple
+                        (_
+                        :: ({ pexp_desc =
+                                Pexp_constant
+                                  { pconst_desc = Pconst_string _; _ };
+                              _
+                            } as const)
+                        :: _);
+                    _
+                  } )
+            when Longident.head ident.txt = "CamlinternalFormatBasics" ->
+            (* We need to retransform format specification which has been desugared into string. *)
+            const
+          | Pexp_poly (expr, _) ->
+            (* We also have to remove poly extra that cause unexpected "!poly!"
+               to be printed in generated code. This happens when you try
+               to extract the body of a method. *)
+            expr
+          | _ -> Ast_mapper.default_mapper.expr mapper expr)
+    }
+  in
+  mapper.expr mapper expr |> Parsetree_utils.expr_remove_merlin_attributes
+
+module Gen = struct
+  let unit = Longident.Lident "()" |> Location.mknoloc
+
+  (* Generates [let name = body]. *)
+  let toplevel_let ~name ~body =
+    let open Ast_helper in
+    let pattern = Pat.mk (Ppat_var { txt = name; loc = Location.none }) in
+    let body = clean_up_for_printing body in
+    Str.value Nonrecursive [ Vb.mk pattern body ]
+
+  (* Generates [let name () = body]. *)
+  let let_unit_toplevel ~name ~body =
+    let open Ast_helper in
+    let unit_param =
+      { Parsetree.pparam_loc = Location.none;
+        pparam_desc = Pparam_val (Nolabel, None, Pat.construct unit None)
+      }
+    in
+    let body = Exp.function_ [ unit_param ] None (Pfunction_body body) in
+    toplevel_let ~name ~body
+
+  (* Generates [let name params = body]. *)
+  let toplevel_function params ~name ~body =
+    let open Ast_helper in
+    let params =
+      List.map
+        ~f:(fun param ->
+          let pattern =
+            Pat.construct
+              (Location.mknoloc (Longident.Lident (Path.last param)))
+              None
+          in
+          { Parsetree.pparam_loc = Location.none;
+            pparam_desc = Pparam_val (Nolabel, None, pattern)
+          })
+        params
+    in
+    let body = Exp.function_ params None (Pfunction_body body) in
+    toplevel_let ~name ~body
+
+  let ident ~name =
+    Longident.Lident name |> Location.mknoloc |> Ast_helper.Exp.ident
+
+  let fun_apply params ~name =
+    let open Ast_helper in
+    let params =
+      List.map ~f:(fun p -> (Asttypes.Nolabel, clean_up_for_printing p)) params
+    in
+    Exp.apply (ident ~name) params
+
+  let fun_apply_unit = fun_apply [ Ast_helper.Exp.ident unit ]
+
+  let fun_apply_params params =
+    params
+    |> List.map ~f:(fun param -> ident ~name:(Path.name param))
+    |> fun_apply
+end
+
+let source_sub_loc src loc =
```

Comment:
For me, that names implies that the resutl is a `loc`. But the result is a fragment of the source corresponding to the given loc. Maybe `extract_source_around_loc` ?